### PR TITLE
File#extname

### PR DESCRIPTION
**Pull request by gogainda**
Implementation of File#extname in rust https://github.com/danielpclark/faster_path/issues/30

### CODE REVIEW

**Comment by gogainda** [comment]
```
bench_rust_extname       0.230018        0.232702        0.242255        0.237788        0.238273
bench_ruby_extname       0.367709        0.338551        0.340032        0.336541        0.338120
```

File: src/extname.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        } else {
+            break;
+        }
+    }
+
+    if (separator_index >= dot_index) {
+        return CString::new("").unwrap().into_raw();
+    }
+
+    if (reversed_ext.len() == 1) {
```
Comment (mble) [inline src/extname.rs]: Could we do:
```rust
if (separator_index >= dot_index || reversed_ext.len() == 1) {
  // ...
}
```

Comment (gogainda) [inline src/extname.rs]: correct

File: src/extname.rs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    let r_str = str::from_utf8(c_str.to_bytes()).unwrap_or("");
+
+    let mut chars_iter = r_str.char_indices().rev();
+
+    let mut reversed_ext: Vec<u8> = Vec::new();
+    let mut dot_index = 0;
+    let mut separator_index = 0;
+    loop {
+        let char = chars_iter.next();
+        if (char.is_some()) {
```
Comment (glebm) [inline src/extname.rs]: A newbie question, can we use pattern matching instead?

```rust
match chars_iter.next() {
  Some(char) => { ... }, // avoid explicit unwrapping
  None => { break }
}
```

Comment (gogainda) [inline src/extname.rs]: sure we can

Comment (gogainda) [inline src/extname.rs]: strange, but I am gettin worse results, leaving as it is

Comment (gogainda) [inline src/extname.rs]: actually it is pitifull that I have to write ugly code to make it fast, what the point to use rust  then - it is an open question

Comment (glebm) [inline src/extname.rs]: @steveklabnik is pattern-matching not a zero-cost abstraction?

**Comment by danielpclark** [comment]
@gogainda   New policy.  Our methods can't blow up when they get handed a `nil`.  Add a `nil` test and change the `assert!(!string.is_null());` in the unsafe block to:

```rust
if string.is_null() {
  return string;
}
```

**Comment by danielpclark** [comment]
Actually I've changed my mind... I found a way to make the code way faster!

**Comment by gogainda** [comment]
@danielpclark done

**Comment by danielpclark** [comment]
This needs more tests comparing `Pathname.extname` and `File.extname` against each other. Like the chop_basename A,B,C tests.

**Comment by gogainda** [comment]
You mean to add more benchmarks?

**Comment by danielpclark** [comment]
No.  A bunch of `assert_equal File.extname(str), Pathname.extname(str)` tests

File: src/extname.rs (hunk 2 above)
Comment (steveklabnik) [inline src/extname.rs]: It should be. I can't imagine why this generates any different code, unless there's a bug somewhere. How was performance being tested?

Comment (gogainda) [inline src/extname.rs]: rack bench

**Comment by gogainda** [comment]
@danielpclark Updated, rebased

**Comment by danielpclark** [comment]
What does the `unwrap().1` and `unwrap().0` do?  I've never seen `.1` and `.0` used before.

File: test/extname_test.rb [hunk 3]
```diff
@@ -0,0 +1,54 @@
+require 'test_helper'
+
+class ExtnameTest < Minitest::Test
+  def test_nil_inputs
+    assert_nil FasterPath.chop_basename(nil)
```
Comment (danielpclark) [inline test/extname_test.rb]: This needs to be `FasterPath.extname(nil)`

**Comment by gogainda** [comment]
@danielpclark Get nth result

**Comment by gogainda** [comment]
In my case it was tuple

File: test/extname_test.rb (hunk 3 above)
Comment (gogainda) [inline test/extname_test.rb]: correct

**Comment by gogainda** [comment]
@danielpclark done

**Pull request merged**